src/states/joining_peer.rs

```diff
@@ -373,30 +567,78 @@ impl Base for JoiningPeer {
     }
 
     fn verify_message(&self, msg: &Message) -> Result<bool> {
-        match (&msg.variant, &self.join_type) {
-            (Variant::NodeApproval(_), JoinType::Relocate(payload)) => {
+        let join_type = match &self.stage {
+            Stage::Bootstrapping(_) => None,
+            Stage::Joining(stage) => Some(&stage.join_type),
+        };
+
+        match (&msg.variant, join_type) {
+            (Variant::NodeApproval(_), Some(JoinType::Relocate(payload))) => {
                 let details = payload.relocate_details();
                 let key_info = &details.destination_key_info;
                 self.verify_message_full(msg, Some(key_info))
             }
-            (Variant::NodeApproval(_), JoinType::First { .. }) => {
+            (Variant::NodeApproval(_), Some(JoinType::First { .. })) => {
                 // We don't have any trusted keys to verify this message, but we still need to
                 // handle it.
                 Ok(true)
             }
-            (Variant::BootstrapResponse(BootstrapResponse::Join(_)), _)
-            | (Variant::Bounce { .. }, _) => self.verify_message_full(msg, None),
+            (Variant::BootstrapResponse(_), _) | (Variant::Bounce { .. }, _) => {
+                self.verify_message_full(msg, None)
+            }
             _ => unreachable!("unexpected message to verify: {:?}", msg),
         }
     }
 }
 
 impl Display for JoiningPeer {
     fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
-        write!(formatter, "JoiningPeer({})", self.name())
+        write!(formatter, "JoiningPeer({}", self.name())?;
+
+        if let Stage::Joining(stage) = &self.stage {
+            write!(formatter, "({:b})", stage.elders_info.prefix())?
+        }
+
+        write!(formatter, ")")
     }
 }
 
+// Stage of joining the network.
+enum Stage {
+    Bootstrapping(BootstrappingStage),
+    Joining(JoiningStage),
+}
+
+impl Stage {
+    fn new(relocate_details: Option<SignedRelocateDetails>) -> Self {
+        Self::Bootstrapping(BootstrappingStage {
+            pending_requests: Default::default(),
+            timeout_tokens: Default::default(),
+            relocate_details,
+        })
+    }
+
+    fn is_bootstrapping(&self) -> bool {
+        matches!(self, Self::Bootstrapping(_))
+    }
+
+    fn is_joining(&self) -> bool {
+        matches!(self, Self::Joining(_))
+    }
+}
+
+struct BootstrappingStage {
+    // Using `FxHashSet` for deterministic iteration order.
+    pending_requests: FxHashSet<SocketAddr>,
```

Comment:
this `pending_requests` is more like: peers that we have pending bootstrap requests to 
right?

Reply:
Yes, but I find the name descriptive enough, given it scope. Calling it `peers_with_pending_requests` or something like that would be unnecessarily verbose.